consensus/hotstuff/sigvalidator/combined_sig_validator.go

```diff
@@ -0,0 +1,27 @@
+package sigvalidator
+
+import (
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// CombinedSigValidator implements the SigValidator interface and owns the knowledge of how to
+// decode the signer and signature fields. It converts the vote/block into general message and
+// signatures for the underneath stateful verifier to do pure signature verify.
+// It's used by consensus cluster, each signer signs two sigs: staking sig and threshold sig,
+// and combine them together.
+// The validation will first split the combined sig into two parts, then validate each sig part.
+// A stateful sigVerifier is needed to verify and accumulated the threshold sig shares.
+type CombinedSigValidator struct {
```

Comment:
are you missing `ValidateQC` here?

Reply:
ValidateQC could be implemented stateless, whereas this validator is stateful. I wanted to implement the stateless ValidateQC in [`SigValidator`](https://github.com/onflow/flow-go/pull/1089/files#diff-f9b5ea84da12dfb96de66c2b0852b566bce73fac11738ce0656d4017ededb266R25)

Reply:
but who will implement `ValidateQC`? Struct that embeds `CombinedSigValidator`?

Reply:
[The `Validator` module](https://github.com/onflow/flow-go/blob/master/consensus/hotstuff/validator/validator.go#L37).

The current Validator implements both the hotstuff checks and signature checks statelessly. 

We will change it so that:
1. Validator will do hotstuff checks, which are stateless, including QC validation. This will be used by EventHandler.
2. SigValidator will implement the signature checks, which are stateful, and will be used by VoteAggregator/VoteCollector.

Reply:
Actually you are right. 

The SigValidator doesn't have ValidateQC, and therefore the stateful implementation: CombinedSigValidator and SingleSigValidator don't need to implement it.

I created another QCValidator instance for validating QC. And I [explained the difference in the PR](https://github.com/onflow/flow-go/pull/1122)